### PR TITLE
Make bsalloc propagate allocation failures from mmap-alloc

### LINKED ISSUE #4: bsalloc: Support allocation failures.
Currently, `bsalloc` is written to assume that memory allocation (in particular, from `mmap`) can never fail. Instead, it should gracefully handle allocation failure and propagate the error up to the user.

### PR DESCRIPTION
**Pull request by whentze**
Instead of always `expect`-ing a successful allocation from mmap-alloc,
propagate AllocErrs up to the user. 

Fixes #4

### CODE REVIEW

**Comment by ezrosent** [comment]
Thanks for the PR. This is a good start, I just have a couple comments:

1. Could you rephrase the commit message to something like "bsalloc: Propagate allocation failures from mmap-alloc" for consistency with other project contributions. 
2. You currently only account for allocation failures for the "large" allocations that hit bsalloc. I think it is worth making a corresponding change in the macro-generated allocators as well. This means modifying the `unsafe fn alloc` declaration in the `sized_cache` macro, and getting read of the `.expect` call to mmap. Let me know if you'd like more direct pointers to that code.

**Comment by whentze** [comment]
You're right, I missed that!

It should be fine now. I changed the message as well.

**Comment by joshlf** [comment]
Hi @whentze , thanks for the PR! I'll let @ezrosent sign off on these changes, but could you add an entry in bsalloc's `CHANGELOG.md` to reflect these changes? Thanks!

If you want ideas, something like the following would work:

```text
### Changed
- Changed allocation routines so that allocation failure is properly reported
```
